akka-actor/src/main/scala/akka/dispatch/affinity/AffinityPool.scala

```diff
@@ -0,0 +1,305 @@
+  case object ShuttingDown extends PoolState {
+    override def order: Int = 1
+  }
+
+  // does not accept new tasks, does not process tasks in queue
+  case object ShutDown extends PoolState {
+    override def order: Int = 2
+  }
+
+  // all threads have been stopped, does not process tasks and does not accept new ones
+  case object Terminated extends PoolState {
+    override def order: Int = 3
+  }
+}
+
+final class BoundedTaskQueue(capacity: Int) extends AbstractBoundedNodeQueue[Runnable](capacity)
+
+class AffinityPoolConfigurator(config: Config, prerequisites: DispatcherPrerequisites) extends ExecutorServiceConfigurator(config, prerequisites) {
+
+  sealed trait CPUAffinityStrategy {
+    def javaStrat: AffinityStrategies
+  }
+
+  case object NoAffinityStrategy extends CPUAffinityStrategy {
+    override val javaStrat: AffinityStrategies = AffinityStrategies.ANY
+  }
+  case object SameCoreStrategy extends CPUAffinityStrategy {
+    override val javaStrat: AffinityStrategies = AffinityStrategies.SAME_CORE
+  }
+  case object SameSocketStrategy extends CPUAffinityStrategy {
+    override val javaStrat: AffinityStrategies = AffinityStrategies.SAME_SOCKET
+  }
+  case object DifferentCoreStrategy extends CPUAffinityStrategy {
+    override val javaStrat: AffinityStrategies = AffinityStrategies.DIFFERENT_CORE
+  }
+  case object DifferentSocketStrategy extends CPUAffinityStrategy {
+    override val javaStrat: AffinityStrategies = AffinityStrategies.DIFFERENT_SOCKET
+  }
+
+  private def toAffinityStrategy(s: String): CPUAffinityStrategy = s match {
+    case "any"              ⇒ NoAffinityStrategy
+    case "same-core"        ⇒ SameCoreStrategy
+    case "same-socket"      ⇒ SameSocketStrategy
+    case "different-core"   ⇒ DifferentCoreStrategy
+    case "different-socket" ⇒ DifferentSocketStrategy
+    case x                  ⇒ throw new IllegalArgumentException("[%s] is not a valid cpu affinity strategy [any, same-core, same-socket, different-core, different-socket]!" format x)
+
+  }
+
+  private def toWaitingStrategy(s: String): WaitingStrategy = s match {
+    case "sleep"     ⇒ sleepWaitingStrategy
+    case "yield"     ⇒ yieldWaitingStrategy
+    case "busy-spin" ⇒ busySpinWaitingStrategy
+    case x           ⇒ throw new IllegalArgumentException("[%s] is not a valid waiting strategy[sleep, yield, same-socket, busy-spin]!" format x)
+
+  }
+
+  private val poolSize = ThreadPoolConfig.scaledPoolSize(
+    config.getInt("parallelism-min"),
+    config.getDouble("parallelism-factor"),
+    config.getInt("parallelism-max"))
+  private val affinityGroupSize = config.getInt("affinity-group-size")
+  private val strategies = config.getStringList("cpu-affinity-strategies").map(toAffinityStrategy)
+  private val waitingStrat = toWaitingStrategy(config.getString("worker-waiting-strategy"))
+  private val tf = new AffinityThreadFactory("affinity-thread-fact", strategies.map(_.javaStrat): _*)
```

Comment:
It also confirms my observation that we should not allow more threads than available cores. Perhaps we simply cap it to at most that number of threads. I think the library provide some way to find out how many cores that are available for affinity binding.

I tried AffinityThreadFactory with `DIFFERENT_CORE, ANY` on the 48 core box (6 socket x 8 cores). It looks alright but seems like it sticks better to the same socket when using `SAME_CORE, SAME_SOCKET, ANY`:

I think we prefer same socket as much as possible.

@peter-lawrey, any thoughts about socket aspect ?